cmake/CMakeLists.txt

```diff
@@ -1369,7 +1369,8 @@ function(onnxruntime_configure_target target_name)
   #if(WIN32 AND onnxruntime_ENABLE_STATIC_ANALYSIS)
   #  set_target_properties(${target_name} PROPERTIES VS_USER_PROPS ${PROJECT_SOURCE_DIR}/EnableVisualStudioCodeAnalysis.props)
   #endif()
-  target_include_directories(${target_name} PRIVATE ${CMAKE_CURRENT_BINARY_DIR} ${ONNXRUNTIME_ROOT} ${abseil_cpp_SOURCE_DIR})
+  target_include_directories(${target_name} PRIVATE ${CMAKE_CURRENT_BINARY_DIR} ${ONNXRUNTIME_ROOT})
+  target_include_directories(${target_name} SYSTEM PRIVATE ${abseil_cpp_SOURCE_DIR})
```

Comment:
This is a new feature. I'm not sure we are ready for it or not. It does more things than ignoring the warning. And it might require updating cmake version to 3.25.

Reply:
I don't think it is new, because the `SYSTEM` is available at least from [cmake v3.0](https://cmake.org/cmake/help/v3.0/command/target_include_directories.html)

Reply:
Got. You are right, this syntax is available since a very old cmake version. 

But, this property does more than just ignoring the warnings. See the discussion at https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70129 

"The corollary is that you shouldn't use it unless you really need to change the search order for system directories!"

When the build flag was added to gcc, it was only for who developed Linux. The operating system! Nowadays we rarely see people using it.  We should follow his professional opinion.

Reply:
Yeah, It has side effect. But since it hasn't fail the CI pipelines in some nasty way, so I will assume the impact will be minimal. 

I have built-from-source clang that can trigger the same error. And please note this is the only clean way to fix it. I have try various way to surpress the error (warning).
1. -Wno-error=deprecated-builtins
   a) this works well for the new version of clang
   b) this trigger unknown warning options error on 5.3.2 hip-clang
2. -Wno-deprecated-builtins for HIP translation units.
   a) option only available with new clang
   b) to use it, we need to introduce `check_hip_compiler_flag` which will depend on HIP_COMPILER, so we need to configure hip toolchains much earlier. This will need slight structural change on our `CMakeLists.txt`

Reply:
In the above GCC link, Jonathan Wakely gave a solution:

> #pragma GCC diagnostic push
> #pragma GCC diagnostic ignored "-Wdeprecated-declarations"
> #pragma GCC diagnostic ignored "-Wunused-local-typedefs"
> #include <third-party-library/header.h>
> #pragma GCC diagnostic pop

It is exactly what we were doing in the past.

https://github.com/microsoft/onnxruntime/blob/main/include/onnxruntime/core/common/eigen_common_wrapper.h

Reply:
With 
```cpp
#ifdef __clang__
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-builtins"
#endif

#include <absl/container/inlined_vector.h>

#ifdef __clang__
#pragma clang diagnostic pop
#endif
```

On new hipclang, the warning is correctly surpressed. But with hipclang of rocm 5.3 and older, 
```console
onnxruntime/include/onnxruntime/core/framework/tensor_shape.h:27:34: error: unknown warning group '-Wdeprecated-builtins', ignored [-Werror,-Wunknown-warning-option]
```

Reply:
```c++
#include "onnxruntime_config.h"
#ifdef HAS_DEPRECATED_BUILTINS
#ifdef __clang__
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-builtins"
#endif
#endif


#include <absl/container/inlined_vector.h>
```

For example,  in eigen_common_wrapper.h we have:

```c++
#ifdef HAS_DEPRECATED_COPY
#pragma GCC diagnostic ignored "-Wdeprecated-copy"
#endif

```

Reply:
We are using hipclang and gcc simutaneously and hipclang is used for hip language exclusively but `HAS_DEPRECATED_BUILTINS` is checked from cxx compiler. So `HAS_DEPRECATED_BUILTINS` only indicate the gcc has "-Wdeprecated-builtins", not hipclang

Reply:
Does ROCM or hipclang has a macro to indicate its version? 

For GCC we have:
```
#if __GNUC__ >= 6
#pragma GCC diagnostic ignored "-Wignored-attributes"
#endif
```

We may use the similar things for clang and hipclang.